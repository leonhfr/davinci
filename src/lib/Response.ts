// Packages.
import * as Wittgenstein from '@scenicroutes/wittgenstein';

// Internal.
import { Formatter } from './Formatter';
import * as Types from '../types';

// Code.
export class Response {
  readonly page: number;
  readonly pages: number;
  readonly perpage: number;
  readonly total: number;
  readonly photo: Array<Wittgenstein.Photo>;

  static create(input: unknown): Response | Error {
    input = Formatter.formatResponse(input);

    if (isResponse(input)) {
      const { photo, ...other } = input;
      const safePhoto: Array<Wittgenstein.Photo> = [];

      for (const photoItem of photo) {
        const maybePhoto = Wittgenstein.Photo.create(
          Formatter.formatPhoto(photoItem)
        );

        if (maybePhoto instanceof Error) {
          return maybePhoto;
        }
        safePhoto.push(maybePhoto);
      }

      return new Response({ ...other, photo: safePhoto });
    }

    const errMsg = isSafeResponse(input).errMsg;
    return new Error(`ValidationError: ${errMsg}`);
  }

  private constructor(input: CreateResponseInput) {
    this.page = input.page;
    this.pages = input.pages;
    this.perpage = input.perpage;
    this.total = input.total;
    this.photo = input.photo;
  }
}

export const RESPONSE_PROPS = ['page', 'pages', 'perpage', 'total', 'photo'];

export const isResponse = (input: unknown): input is CreateResponseInput => {
  return isSafeResponse(input).isSafe;
};

export const isSafeResponse = (input: unknown): Types.IsSafe => {
  if (typeof input !== 'object') {
    return {
      isSafe: false,
      errMsg: `Expected type of input to be object, got ${typeof input}.`,
    };
  }

  if (input === null) {
    return {
      isSafe: false,
      errMsg: `Expected input not to be null.`,
    };
  }

  for (const key of RESPONSE_PROPS) {
    if (!input.hasOwnProperty(key)) {
      return {
        isSafe: false,
        errMsg: `Expected all input properties to be set, missing ${key}.`,
      };
    }
  }

  const { page, pages, perpage, total } = input as {
    page: unknown;
    pages: unknown;
    perpage: unknown;
    total: unknown;
  };

  const numbers = { page, pages, perpage, total };

  for (const [key, value] of Object.entries(numbers)) {
    if (typeof value !== 'number' || value === NaN) {
      return {
        isSafe: false,
        errMsg: `Expected type of input.${key} to be number, got ${typeof value}.`,
      };
    }
  }

  return {
    isSafe: true,
    errMsg: '',
  };
};

export interface CreateResponseInput {
  readonly page: number;
  readonly pages: number;
  readonly perpage: number;
  readonly total: number;
  readonly photo: Array<Wittgenstein.Photo>;
}
