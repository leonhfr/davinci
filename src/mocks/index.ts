// Internal.
import { CreateResponseInput } from '../lib/Response';
import * as Types from '../types';

// Code.
export const createMockResponseInput = (
  response?: any
): CreateResponseInput => {
  const _response = response || {};
  return {
    page: _response.page || 1,
    pages: _response.pages || 25225,
    perpage: _response.perpage || 2,
    total: _response.total || 50450,
    photo: _response.photo || [],
  };
};

export const createMockResponse = (response?: any) => {
  const _response = response || {};
  return {
    page: _response.page || 1,
    pages: _response.pages || '25225',
    perpage: _response.perpage || 2,
    total: _response.total || '50450',
    photo: _response.photo || [],
  };
};

export const createMockPhoto = (photo?: any) => {
  const _photo = photo || {};
  return {
    id: _photo.id || '32457360587',
    owner: _photo.owner || '41710899@N08',
    secret: _photo.secret || '69b27ceb18',
    server: _photo.server || '4898',
    farm: _photo.farm || 5,
    title: _photo.title || 'Arc de Triomf',
    ispublic: 1,
    isfriend: 0,
    isfamily: 0,
    description: _photo.description || {
      _content:
        'The Arc de Triomf was built as the entrance of the 1888 Barcelona Universal Exposition. It was designed by architect Josep Vilaseca i Casanovas in the Neo-Mudéjar style.',
    },
    ownername: _photo.ownername || 'jdf_92',
    views: _photo.views || '16',
    tags:
      _photo.tags ||
      'spain barcelona arcdetriomf 1888barcelonaworldfair neomudéjar 1888barcelonauniversalexposition catelonia',
    latitude: _photo.latitude || '41.390791',
    longitude: _photo.longitude || '2.18114',
    accuracy: 16,
    context: _photo.context || 0,
    place_id: 'iFsnXNZTUriu3usbkw',
    woeid: '20220092',
    geo_is_family: 0,
    geo_is_friend: 0,
    geo_is_contact: 0,
    geo_is_public: 1,
  };
};

export const photoProperties: Types.PhotoProperties = {
  zoneId: '4ab7068b-6c6c-46d2-8009-1d7d1ab35a3b',
  inside: false,
};

export const apiKey = 'test_api_key';

export const searchOptions: Types.SearchOptions = {
  bbox: [2.18, 41.4, 2.19, 41.5],
  min_upload_date: 1546300800,
  max_upload_date: 1546387200,
  tags: ['foo', 'bar'],
  page: undefined,
};
