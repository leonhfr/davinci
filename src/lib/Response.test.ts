// Internal.
import { Response, RESPONSE_PROPS, isResponse } from './Response';
import * as Mocks from '../mocks';

// Code.
describe('Response', () => {
  describe('Class', () => {
    it('should be defined', () => {
      expect(Response).toBeDefined();
    });

    describe('create', () => {
      it('should fail to create a new instance - invalid input', () => {
        expect(Response.create(null)).toMatchSnapshot();
      });
      it('should create a new instance', () => {
        expect(
          Response.create(Mocks.createMockResponseInput())
        ).toMatchSnapshot();
      });
      it('should return errors when the photos are not correctly formatted', () => {
        expect(
          Response.create({
            ...Mocks.createMockResponse(),
            photo: [Mocks.createMockPhoto()],
          })
        ).toMatchSnapshot();
      });
      it('should create a new instance when the photos are correctly formatted', () => {
        expect(
          Response.create(
            {
              ...Mocks.createMockResponse(),
              photo: [Mocks.createMockPhoto()],
            },
            { zoneId: '4ab7068b-6c6c-46d2-8009-1d7d1ab35a3b', inside: false }
          )
        ).toMatchSnapshot();
      });
    });
  });

  describe('Validation', () => {
    it('should be defined', () => {
      expect(RESPONSE_PROPS).toBeDefined();
    });
    it('should match the expected value', () => {
      expect(RESPONSE_PROPS).toMatchSnapshot();
    });
    it('should be up to date', () => {
      expect(RESPONSE_PROPS).toEqual(
        Object.keys(Response.create(Mocks.createMockResponseInput()))
      );
    });
  });

  describe('isResponse', () => {
    it('should be defined', () => {
      expect(isResponse).toBeDefined();
    });
    it('should handle an invalid input - wrong type', () => {
      expect(isResponse('__FAIL__')).toBe(false);
    });
    it('should handle an invalid input - null type', () => {
      expect(isResponse(null)).toBe(false);
    });
    it('should handle an invalid input - missing props', () => {
      expect(isResponse({})).toBe(false);
    });
    it('should handle an invalid input - numbers', () => {
      expect(
        isResponse({
          ...Mocks.createMockResponseInput(),
          page: '__FAIL__',
        })
      ).toBe(false);
    });
  });
});
