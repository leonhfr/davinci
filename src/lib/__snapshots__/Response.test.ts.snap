// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response Class create should create a new instance 1`] = `
Response {
  "page": 1,
  "pages": 25225,
  "perpage": 2,
  "photo": Array [],
  "total": 50450,
}
`;

exports[`Response Class create should create a new instance when the photos are correctly formatted 1`] = `
Response {
  "page": 1,
  "pages": 25225,
  "perpage": 2,
  "photo": Array [
    Photo {
      "context": 0,
      "description": "The Arc de Triomf was built as the entrance of the 1888 Barcelona Universal Exposition. It was designed by architect Josep Vilaseca i Casanovas in the Neo-Mudéjar style.",
      "farm": 5,
      "id": "32457360587",
      "inside": false,
      "latitude": 41.390791,
      "longitude": 2.18114,
      "owner": "41710899@N08",
      "ownername": "jdf_92",
      "secret": "69b27ceb18",
      "server": "4898",
      "tags": Array [
        "spain",
        "barcelona",
        "arcdetriomf",
        "1888barcelonaworldfair",
        "neomudéjar",
        "1888barcelonauniversalexposition",
        "catelonia",
      ],
      "title": "Arc de Triomf",
      "views": 16,
      "zoneId": "4ab7068b-6c6c-46d2-8009-1d7d1ab35a3b",
    },
  ],
  "total": 50450,
}
`;

exports[`Response Class create should fail to create a new instance - invalid input 1`] = `[Error: ValidationError: Expected input not to be null.]`;

exports[`Response Class create should return errors when the photos are not correctly formatted 1`] = `[Error: [Photo Validation] ValidationError: Expected all input properties to be set, missing zoneId.]`;

exports[`Response Validation should match the expected value 1`] = `
Array [
  "page",
  "pages",
  "perpage",
  "total",
  "photo",
]
`;
